tests/test_quorum.py

```diff
@@ -0,0 +1,473 @@
+import unittest
+
+from typing import List, Set, Tuple
+
+from patroni.quorum import QuorumStateResolver, QuorumError
+
+
+class QuorumTest(unittest.TestCase):
+
+    def check_state_transitions(self, leader: str, quorum: int, voters: Set[str], numsync: int, sync: Set[str],
+                                numsync_confirmed: int, active: Set[str], sync_wanted: int, leader_wanted: str,
+                                expected: List[Tuple[str, str, int, Set[str]]]) -> None:
+        kwargs = {
+            'leader': leader, 'quorum': quorum, 'voters': voters,
+            'numsync': numsync, 'sync': sync, 'numsync_confirmed': numsync_confirmed,
+            'active': active, 'sync_wanted': sync_wanted, 'leader_wanted': leader_wanted
+        }
+        result = list(QuorumStateResolver(**kwargs))
+        self.assertEqual(result, expected)
+
+        # also check interrupted transitions
+        if len(result) > 0 and result[0][0] != 'restart' and kwargs['leader'] == result[0][1]:
+            if result[0][0] == 'sync':
+                kwargs.update(numsync=result[0][2], sync=result[0][3])
+            else:
+                kwargs.update(leader=result[0][1], quorum=result[0][2], voters=result[0][3])
+            kwargs['expected'] = expected[1:]
+            self.check_state_transitions(**kwargs)
+
+    def test_1111(self):
```

Comment:
What is the convention on these test names that have trailing numbers?

Reply:
Very good question :)
These names are from the original PR: https://github.com/zalando/patroni/pull/672/files#diff-77b9ff9cf878f34d15fd5bada8be62ca3475ab49e9d6f15db4261728837be972R6

I think I'll rename them one day.